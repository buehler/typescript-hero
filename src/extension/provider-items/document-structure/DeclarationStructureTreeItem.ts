import { ExtensionContext, TreeItemCollapsibleState } from 'vscode';
import { CompletionItemKind } from 'vscode-languageserver-types';

import {
    ClassDeclaration,
    Declaration,
    FunctionDeclaration,
    InterfaceDeclaration,
    MethodDeclaration,
    PropertyDeclaration,
    VariableDeclaration,
} from '../../../common/ts-parsing/declarations';
import { stringTemplate } from '../../../common/utilities/StringTemplate';
import { BaseStructureTreeItem } from './BaseStructureTreeItem';

const fileTemplate = stringTemplate`./src/extension/assets/icons/declarations/${0}.svg`;

/**
 * Function to calculate the displayed name of the declaration structure item.
 * 
 * @param {Declaration} declaration 
 * @returns {string} 
 */
function getDeclarationLabel(declaration: Declaration): string {
    if (
        declaration instanceof FunctionDeclaration ||
        declaration instanceof MethodDeclaration
    ) {
        const params = declaration.parameters.map(p => p.name + (p.type ? `: ${p.type}` : '')).join(', ');
        return `${declaration.name}(${params})${declaration.type ? `: ${declaration.type}` : ''}`;
    }

    if (declaration instanceof PropertyDeclaration) {
        return declaration.name + (declaration.type ? `: ${declaration.type}` : '');
    }

    if (
        declaration instanceof ClassDeclaration ||
        declaration instanceof InterfaceDeclaration
    ) {
        return declaration.name + (declaration.typeParameters ? `<${declaration.typeParameters.join(', ')}>` : '');
    }

    return declaration.name;
}

/**
 * Structure item that represents a typescript declaration of any way.
 * 
 * @export
 * @class DeclarationStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export class DeclarationStructureTreeItem extends BaseStructureTreeItem {
    public get iconPath(): string | undefined {
        switch (this.declaration.itemKind) {
            case CompletionItemKind.Class:
            case CompletionItemKind.Keyword:
                return this.context.asAbsolutePath(fileTemplate('class'));
            case CompletionItemKind.Interface:
                return this.context.asAbsolutePath(fileTemplate('interface'));
            case CompletionItemKind.Enum:
                return this.context.asAbsolutePath(fileTemplate('enum'));
            case CompletionItemKind.Function:
            case CompletionItemKind.Method:
                return this.context.asAbsolutePath(fileTemplate('callable'));
            case CompletionItemKind.Module:
                return this.context.asAbsolutePath(fileTemplate('module'));
            case CompletionItemKind.Property:
                return this.context.asAbsolutePath(fileTemplate('property'));
            default:
                break;
        }

        if (this.declaration.itemKind === CompletionItemKind.Variable) {
            return (this.declaration as VariableDeclaration).isConst ?
                this.context.asAbsolutePath(fileTemplate('const')) :
                this.context.asAbsolutePath(fileTemplate('variable'));
        }

        return this.context.asAbsolutePath(fileTemplate('default'));
    }

    constructor(public declaration: Declaration, private context: ExtensionContext) {
        super(getDeclarationLabel(declaration));

        if (
            declaration instanceof ClassDeclaration ||
            declaration instanceof InterfaceDeclaration
        ) {
            this.collapsibleState = TreeItemCollapsibleState.Expanded;
        }
    }

    public getChildren(): BaseStructureTreeItem[] {
        if (
            this.declaration instanceof ClassDeclaration ||
            this.declaration instanceof InterfaceDeclaration
        ) {
            return [
                ...this.declaration.properties.map(p => new DeclarationStructureTreeItem(p, this.context)),
                ...this.declaration.methods.map(m => new DeclarationStructureTreeItem(m, this.context)),
            ];
        }
        return [];
    }
}

