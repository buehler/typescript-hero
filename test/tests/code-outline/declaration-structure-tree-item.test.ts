import { relative } from 'path';
import {
  ClassDeclaration,
  EnumDeclaration,
  FunctionDeclaration,
  GetterDeclaration,
  InterfaceDeclaration,
  MethodDeclaration,
  ModuleDeclaration,
  PropertyDeclaration,
  TypeAliasDeclaration,
  VariableDeclaration,
} from 'typescript-parser';
import { ExtensionContext } from 'vscode';

import DeclarationStructureTreeItem from '../../../src/code-outline/declaration-structure-tree-item';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';
import { expect } from '../setup';

declare global {
  namespace Chai {
    interface Assertion {
      matchSnapshot(): Assertion;
    }
  }
}

describe('DeclarationStructureTreeItem', () => {

  let context: ExtensionContext;

  before(() => {
    context = ioc.get<ExtensionContext>(iocSymbols.extensionContext);
  });

  it('should create a tree item', () => {
    const declaration = new ClassDeclaration('class', true, 0, 100);
    const item = new DeclarationStructureTreeItem(declaration, context);

    expect(item).to.exist;
  });

  const iconPossibilities = [
    {
      name: 'class',
      declaration: new ClassDeclaration('class', true, 0, 100),
    },
    {
      name: 'interface',
      declaration: new InterfaceDeclaration('interface', true, 0, 100),
    },
    {
      name: 'enum',
      declaration: new EnumDeclaration('enum', true, 0, 100),
    },
    {
      name: 'function',
      declaration: new FunctionDeclaration('function', true, 'void', 0, 100),
    },
    {
      name: 'method',
      declaration: new MethodDeclaration('method', false, undefined, 'void', 0, 100),
    },
    {
      name: 'module',
      declaration: new ModuleDeclaration('module', 0, 100),
    },
    {
      name: 'property',
      declaration: new PropertyDeclaration('property', undefined, 'void', 0, 100),
    },
    {
      name: 'variable',
      declaration: new VariableDeclaration('variable', false, true, 'string', 0, 100),
    },
    {
      name: 'const',
      declaration: new VariableDeclaration('const', true, true, 'string', 0, 100),
    },
    {
      name: 'default',
      declaration: new TypeAliasDeclaration('typealias', true, 0, 100),
    },
  ];

  for (const test of iconPossibilities) {
    it(`should return the correct icon path for itemKind "${test.name}"`, () => {
      const item = new DeclarationStructureTreeItem(test.declaration, context);

      const path = relative(global['rootPath'], item.iconPath || '');
      expect(path).to.matchSnapshot();
    });
  }

  it('should return the correct accessor children', () => {
    const declaration = new ClassDeclaration('class', true, 0, 100);
    declaration.accessors.push(new GetterDeclaration('getter', undefined, undefined, false));
    const item = new DeclarationStructureTreeItem(declaration, context);

    expect(item.getChildren().map(c => ({ label: c.label, ctor: c.constructor.name }))).to.matchSnapshot();
  });

  it('should return the correct property children', () => {
    const declaration = new ClassDeclaration('class', true, 0, 100);
    declaration.properties.push(new PropertyDeclaration('property', undefined, undefined));
    const item = new DeclarationStructureTreeItem(declaration, context);

    expect(item.getChildren().map(c => ({ label: c.label, ctor: c.constructor.name }))).to.matchSnapshot();
  });

  it('should return the correct method children', () => {
    const declaration = new ClassDeclaration('class', true, 0, 100);
    declaration.methods.push(new MethodDeclaration('method', false, undefined, undefined));
    const item = new DeclarationStructureTreeItem(declaration, context);

    expect(item.getChildren().map(c => ({ label: c.label, ctor: c.constructor.name }))).to.matchSnapshot();
  });

  it('should not return children on simple declarations', () => {
    const declaration = new VariableDeclaration('variable', false, true, undefined);
    const item = new DeclarationStructureTreeItem(declaration, context);

    expect(item.getChildren()).to.matchSnapshot();
  });

});
