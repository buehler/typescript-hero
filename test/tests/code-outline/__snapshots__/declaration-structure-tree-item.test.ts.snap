// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclarationStructureTreeItem should not return children on simple declarations 1`] = `Array []`;

exports[`DeclarationStructureTreeItem should return the correct accessor children 1`] = `
Array [
  DeclarationStructureTreeItem {
    "collapsibleState": 0,
    "command": Object {
      "arguments": Array [
        GetterDeclaration {
          "end": undefined,
          "isAbstract": false,
          "name": "getter",
          "start": undefined,
          "type": undefined,
          "visibility": undefined,
        },
      ],
      "command": "typescriptHero.codeOutline.gotoNode",
      "title": "Jump to node",
    },
    "context": ContextMock {
      "extensionPath": "",
      "storagePath": "",
      "subscriptions": Array [],
    },
    "declaration": GetterDeclaration {
      "end": undefined,
      "isAbstract": false,
      "name": "getter",
      "start": undefined,
      "type": undefined,
      "visibility": undefined,
    },
    "label": "get() getter",
  },
]
`;

exports[`DeclarationStructureTreeItem should return the correct icon path 1`] = `"./src/assets/icons/declarations/class.svg"`;

exports[`DeclarationStructureTreeItem should return the correct method children 1`] = `
Array [
  DeclarationStructureTreeItem {
    "collapsibleState": 0,
    "command": Object {
      "arguments": Array [
        MethodDeclaration {
          "end": undefined,
          "isAbstract": false,
          "name": "method",
          "parameters": Array [],
          "start": undefined,
          "type": undefined,
          "variables": Array [],
          "visibility": undefined,
        },
      ],
      "command": "typescriptHero.codeOutline.gotoNode",
      "title": "Jump to node",
    },
    "context": ContextMock {
      "extensionPath": "",
      "storagePath": "",
      "subscriptions": Array [],
    },
    "declaration": MethodDeclaration {
      "end": undefined,
      "isAbstract": false,
      "name": "method",
      "parameters": Array [],
      "start": undefined,
      "type": undefined,
      "variables": Array [],
      "visibility": undefined,
    },
    "label": "method()",
  },
]
`;

exports[`DeclarationStructureTreeItem should return the correct property children 1`] = `
Array [
  DeclarationStructureTreeItem {
    "collapsibleState": 0,
    "command": Object {
      "arguments": Array [
        PropertyDeclaration {
          "end": undefined,
          "name": "property",
          "start": undefined,
          "type": undefined,
          "visibility": undefined,
        },
      ],
      "command": "typescriptHero.codeOutline.gotoNode",
      "title": "Jump to node",
    },
    "context": ContextMock {
      "extensionPath": "",
      "storagePath": "",
      "subscriptions": Array [],
    },
    "declaration": PropertyDeclaration {
      "end": undefined,
      "name": "property",
      "start": undefined,
      "type": undefined,
      "visibility": undefined,
    },
    "label": "property",
  },
]
`;
